core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs

```diff
@@ -0,0 +1,243 @@
+use std::{
+    fmt::Debug,
+    sync::{Arc, Mutex},
+};
+
+use anyhow::Context;
+use async_trait::async_trait;
+use tokio::{runtime::Handle, sync::watch};
+use zksync_dal::{ConnectionPool, StorageProcessor};
+use zksync_state::{PostgresStorage, ReadStorage, RocksdbStorage, StateKeeperColumnFamily};
+use zksync_storage::RocksDB;
+
+/// Encapsulates a storage that can produce a [`ReadStorageFactory`] on demand.
+///
+/// This struct's main design purpose is to be able to produce a (cheap) owned [`ReadStorageFactory`]
+/// implementation that can then be used to obtain a short-lived [`ReadStorage`] handle bound to
+/// the respective factory.
+#[derive(Debug, Clone)]
+pub struct StateKeeperStorage<T: ReadStorageFactory> {
+    // FIXME: `Arc<Mutex<_>>` is a bit of a leaky abstraction here as it exists mostly because of
+    // the mutable nature of [`AsyncRocksdbCache`].
+    inner: Arc<Mutex<T>>,
+}
+
+/// Factory that can produce a [`ReadStorage`] implementation on demand.
+#[async_trait]
+pub trait ReadStorageFactory: Clone + Debug + Send + Sync {
+    type ReadStorageImpl<'a>: ReadStorage
+    where
+        Self: 'a;
+
+    async fn access_storage<'a>(
+        &'a self,
+        rt_handle: Handle,
+        stop_receiver: &watch::Receiver<bool>,
+    ) -> Option<Self::ReadStorageImpl<'a>>;
+}
+
+/// A [`ReadStorage`] implementation that uses either [`PostgresStorage`] or [`RocksdbStorage`]
+/// underneath.
+#[derive(Debug)]
+pub enum PgOrRocksdbStorage<'a> {
+    Postgres(PostgresStorage<'a>),
+    RocksdbStorage(RocksdbStorage),
+}
+
+impl<'a> ReadStorage for PgOrRocksdbStorage<'a> {
+    fn read_value(&mut self, key: &zksync_types::StorageKey) -> zksync_types::StorageValue {
+        match self {
+            PgOrRocksdbStorage::Postgres(postgres) => postgres.read_value(key),
```

Comment:
Nit: You can use `Self::*` instead of `PgOrRocksdbStorage::*` in match clauses here and in other `ReadStorage` methods.